### ISSUE TITLE
Regression after LHS change

### ISSUE THREAD

**Issue by josh** [svaarala/duktape#132]
Hey @svaarala,

Just finding lots of stuff today :wink: 

I think I stumbled on some sort of stack or register corruption bug introduced in 863b60dbb427e825ced671142a8b630280a2ecca. Thats what I bisected the problem to. That change is currently unreleased.

``` sh
$  curl https://gist.githubusercontent.com/josh/2f56ec3c4cc25d37d132/raw/52e9e8899691b777eabb5b7a2fb52b160bb1a6a3/babel.js > babel.js
$ ./duk babel.js
TypeError: not number
	duk_api_stack.c:1048
	anon babel.js:12089 strict tailcalled
	anon babel.js:16398 preventsyield
	call  native strict preventsyield
	s babel.js:1 tailcalled
	anon babel.js:2335 preventsyield
	call  native strict preventsyield
	s babel.js:1 tailcalled
	anon babel.js:2534 strict preventsyield
	call  native strict preventsyield
	s babel.js:1 tailcalled
	[...]
error in executing file babel.js
```

The crash happens when it tries to iterate over an object `for (var kw in keywordTypes)` after doing a comma separate assignment to a value with objects, `var _jsxTagStart = {type: "jsxTagStart"}, _jsxTagEnd = {type: "jsxTagEnd"};`. Theres some other required state to be on the stack but I haven't quite figured that part out. If you try to do almost anything between those lines, it magically works. I'm just assuming some register gets cleared out and its somehow fine.

Let me know if theres anything I can add further here. Sorry I wasn't able to boil this down to a simple test case.

**Comment by svaarala** [comment]
No worries, thanks for reporting :)

I'll see if I can reproduce it using babel.js and figure out what's wrong.

**Comment by svaarala** [comment]
Here's some bytecode execution log right before the error:

```
[D] duk_js_executor.c:2072 (duk_js_execute_bytecode): executing bytecode: pc=1605 ins=0x00524080, op=0, valstack_top=518/640  -->  LDREG 2, 329, 0
[D] duk_js_executor.c:2072 (duk_js_execute_bytecode): executing bytecode: pc=1606 ins=0x7e80833e, op=62, valstack_top=518/640  -->  INITENUM 2, 253
[D] duk_js_executor.c:2072 (duk_js_execute_bytecode): executing bytecode: pc=1607 ins=0x80000332, op=50, valstack_top=518/640  -->  JUMP 12 (to pc+13)
[D] duk_js_executor.c:2072 (duk_js_execute_bytecode): executing bytecode: pc=1620 ins=0x00520080, op=0, valstack_top=518/640  -->  LDREG 2, 328, 0
[D] duk_js_executor.c:2072 (duk_js_execute_bytecode): executing bytecode: pc=1621 ins=0x005240c0, op=0, valstack_top=518/640  -->  LDREG 3, 329, 0
[D] duk_js_executor.c:2072 (duk_js_execute_bytecode): executing bytecode: pc=1622 ins=0x0180837e, op=62, valstack_top=518/640  -->  NEXTENUM 2, 3
[D] duk_api_stack.c:1047 (duk_require_number): EXPECTED NUMBER: undefined
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
[D] duk_js_executor.c:1181 (duk__handle_longjmp): -> throw propagated up to entry level, rethrow and exit bytecode executor
TypeError: not number
	duk_api_stack.c:1049
	anon babel.js:12089 strict
	anon babel.js:11628
	anon babel.js:16398 preventsyield
	call  native strict preventsyield
	s babel.js:1
	anon babel.js:1
	anon babel.js:2335 preventsyield
	call  native strict preventsyield
	s babel.js:1
	anon babel.js:1
	[...]
```

**Comment by svaarala** [comment]
One issue I found is a register shuffling issue when emitting INITENUM and NEXTENUM opcodes and the target register(s) need shuffling.  The compiler code was missing a flag to indicate that register in "B slot" was a target, not a source, so it would get shuffled incorrectly.  This only happens when the enum target is assigned to a register outside the directly addressable window (roughly 8 bits), i.e. large functions.

I don't see why this would be related to the LHS change, except indirectly by the change affecting register assignments.

With this fix (and GH-131 fix included) the babel.js file fails at a different point now. Not sure if there's another issue - could you try it out with this branch:

- https://github.com/svaarala/duktape/tree/fix-enum-shuffle-gh132

**Comment by svaarala** [comment]
Managed to create a testcase which reproduces the issue and also reveals a separate issue with NEXTENUM and spilling: NEXTENUM is assuming that the next instruction slot is a jump out of the loop but this is not the case when spilling happens.

**Comment by svaarala** [comment]
Ok, NEXTENUM jump slot fix is also in the branch.  Again, not sure if babel.js now works correctly or not.

**Comment by svaarala** [comment]
The coverage for shuffle behavior is not very good right now because it's only triggered in large functions, so I'm adding a torture option for it: #135.

**Comment by svaarala** [comment]
@josh The above fix branch is now merged to master. It fixes one concrete issue but I'm not sure if babel.js yet works correctly, so I won't close this yet.

**Comment by josh** [comment]
Confirmed. Its passed that one error, but theres still something else up. I'll try to see what I can do to trace down the offending line as well.

Thanks for all your help so far!

**Comment by svaarala** [comment]
@josh Added some debug and re-ran babel.js.  It seems to fail on this:

```
var f;"undefined"!=typeof window?f=window:"undefined"!=typeof global?f=global:"undefined"!=typeof self&&(f=self),f.babel=e()
```

The offending assignment is `f.babel=e()` which is caused by "window" being absent.  I added:

```
var window = {};
```

and now it runs (though does nothing).

**Comment by svaarala** [comment]
The following test included a PUTPROP shuffle fix, issue discovered using shuffle torture testing:

- https://github.com/svaarala/duktape/commit/bb00d0cab92613daa0d9df3fc7ed9d1ebbbfc8b5

**Comment by josh** [comment]
wow! everything is passing here. I have green tests running coffee-script.js, babel.js and uglify.js.

Thanks so much! `master` looks good!

**Comment by svaarala** [comment]
@josh Great! I released v1.1.2 with the shuffle fixes included - it'd be nice if you could confirm that release 1.1.2 works for you same way master does.

**Comment by josh** [comment]
:grin: yes, working great!!!!
